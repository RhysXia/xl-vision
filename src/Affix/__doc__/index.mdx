# 图钉

## 代码演示

:::demo 基本用法

基本用法

```jsx
import React from 'react'
import { Affix } from 'xl-vision'

export default () => {
  return (
    <div>
      <Affix offsetTop={20}>
        <span className='affix'>固定在距离顶部20px的位置</span>
      </Affix>
      <br />
      <Affix offsetBottom={20}>
        <span className='affix'>固定在距离底部20px的位置</span>
      </Affix>
    </div>
  )
}
```

```scss
.affix {
  display: inline-block;
  padding: 8px 16px;
  background-color: #1890ff;
  color: #fff;
}
```

:::

:::demo 状态改变回调

affix 状态改变时的回调

```jsx
import React from 'react'
import { Affix } from 'xl-vision'

export default () => {
  const handleChange = React.useCallback((visible) => {
    console.log(visible)
  }, [])

  return (
    <div>
      <Affix offsetTop={50} offsetBottom={50} onChange={handleChange}>
        <span className='affix'>固定在距离顶部50px和距离底部50px的位置</span>
      </Affix>
    </div>
  )
}
```

```scss
.affix {
  display: inline-block;
  padding: 8px 16px;
  background-color: #1890ff;
  color: #fff;
}
```

:::

:::demo 指定固定在的父元素

通过`getTarget`指定固定的父元素，默认为`window`。

注意，为了性能考虑，此时只会监听`target`的滚动事件，所以，当页面滚动时，就会出现定位错误。

如果有特殊需要，可以调用`Affix`的`updatePosition`方法。

```jsx
import React from 'react'
import { Affix } from 'xl-vision'

export default () => {
  const parentRef = React.useRef()
  const affixRef = React.useRef()

  const getTarget = React.useCallback(() => parentRef.current, [])

  React.useEffect(() => {
    if (!affixRef.current) {
      return
    }

    window.addEventListener('scroll', affixRef.current.updatePosition)
    window.addEventListener('resize', affixRef.current.updatePosition)
    return () => {
      window.removeEventListener('scroll', affixRef.current.updatePosition)
      window.removeEventListener('resize', affixRef.current.updatePosition)
    }
  }, [])

  return (
    <div className='container' ref={parentRef}>
      <div className='parent'>
        <Affix offsetTop={50} getTarget={getTarget} ref={affixRef}>
          <span className='affix'>固定在距离target顶部50px处</span>
        </Affix>
      </div>
    </div>
  )
}
```

```scss
.container {
  position: relative;
  height: 200px;
  overflow-y: scroll;
}

.parent {
  height: 1000px;
  background-color: #eee;
}

.affix {
  display: inline-block;
  padding: 8px 16px;
  background-color: #1890ff;
  color: #fff;
}
```

:::

## Affix API

| 参数 | 说明 | 类型 | 可选值 | 默认值 | 是否必填 |
| --- | --- | --- | --- | --- | --- |
| clsPrefix | 自定义类名前缀 | string | - | - | false |
| children | 被定位元素 | React.ReactNode | - | - | true |
| offsetTop | 距离顶部距离，如果不设置，则不对顶部进行定位 | number | - | - | false |
| offsetBottom | 距离底部距离，如果不设置，则不对底部进行定位 | number | - | - | false |
| getTarget | 定位参考元素 | () => HTMLElement &#x007c; Document &#x007c; Window | - | () => window | false |
| onChange | 定位状态发生变化的回调 | (affix :boolean) => void | - | - | false |
