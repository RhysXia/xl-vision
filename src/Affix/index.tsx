import React from 'react'
import PropTypes from 'prop-types'

export interface AffixProps {
  children: React.ReactElement
  offsetTop?: number
  offsetBottom?: number
  scrollContainer?: () => HTMLElement
  onChange?: (changed: boolean) => void
}

const defaultScrollContainer = () => window

const Affix: React.FunctionComponent<AffixProps> = (props) => {
  const {
    children,
    offsetBottom,
    offsetTop,
    scrollContainer = defaultScrollContainer,
    onChange
  } = props

  return children
}

Affix.displayName = 'Affix'

Affix.propTypes = {
  children: PropTypes.element.isRequired
}

export default Affix
