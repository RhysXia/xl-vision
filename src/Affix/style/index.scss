@use './variables' as *;

$affix: #{$cls-prefix}-affix;

.#{$affix} {
  z-index: $affix-zindex;
}
