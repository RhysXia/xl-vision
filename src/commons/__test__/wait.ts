import { act } from 'react-dom/test-utils'

export default (time: number) => {
  const wait = new Promise<void>(resolve => {
    // console.info(`wait time: ${time}`)
    setTimeout(() => {
      resolve()
    }, time)
  })

  return act(() => wait)
}
