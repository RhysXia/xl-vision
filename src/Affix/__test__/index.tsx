import { mount } from 'enzyme'
import React from 'react'
import { act } from 'react-dom/test-utils'
import wait from '../../../test/wait'
import Affix, { AffixRef } from '..'

describe('Affix', () => {
  const events: Partial<Record<keyof HTMLElementEventMap, (ev: Partial<Event>) => void>> = {}
  const domMock = jest.spyOn(HTMLElement.prototype, 'getBoundingClientRect')
  const classRect: Record<string, DOMRect> = {
    container: {
      top: 0,
      bottom: 100
    } as DOMRect
  }

  beforeAll(() => {
    jest.useRealTimers()

    domMock.mockImplementation(function fn(this: HTMLElement) {
      if (this.className === 'container') {
        return classRect.container
      }
      return (
        classRect.fixed || {
          top: 0,
          bottom: 0,
          width: 10,
          height: 10
        }
      )
    })
  })

  afterAll(() => {
    domMock.mockRestore()
  })

  const movePlaceholder = async (top: number) => {
    classRect.fixed = {
      top,
      bottom: top + 10,
      width: 10,
      height: 10
    } as DOMRect
    if (events.scroll == null) {
      throw new Error('scroll should be set')
    }
    events.scroll({
      type: 'scroll'
    })
    await act(() => wait(20))
  }

  const AffixMock = (props: {
    offsetTop?: number
    offsetBottom?: number
    onChange?: (change: boolean) => void
  }) => {
    const containerRef = React.useRef<HTMLDivElement>()
    const affixRef = React.useRef<AffixRef>(null)

    const getTarget = () => containerRef.current!

    return (
      <div
        className='container'
        ref={(node) => {
          if (node) {
            node.addEventListener = jest
              .fn()
              .mockImplementation(
                (event: keyof HTMLElementEventMap, cb: (ev: Partial<Event>) => void) => {
                  events[event] = cb
                }
              )
          }
          containerRef.current = node!
        }}
      >
        <Affix {...props} getTarget={getTarget} ref={affixRef}>
          <button type='button' className='btn'>
            btn
          </button>
        </Affix>
      </div>
    )
  }

  it('support offsetTop', async () => {
    const handleChange = jest.fn()
    const wrapper = mount(<AffixMock offsetTop={0} onChange={handleChange} />)

    await wait(20)
    expect(handleChange.mock.calls.length).toBe(0)

    expect(wrapper.render()).toMatchSnapshot()

    await movePlaceholder(50)
    expect(handleChange.mock.calls.length).toBe(0)
    expect(wrapper.render()).toMatchSnapshot()

    await movePlaceholder(150)
    expect(handleChange.mock.calls.length).toBe(0)
    expect(wrapper.render()).toMatchSnapshot()

    await movePlaceholder(0)
    expect(handleChange.mock.calls.length).toBe(0)
    expect(wrapper.render()).toMatchSnapshot()

    await movePlaceholder(-50)
    expect(handleChange.mock.calls.length).toBe(1)
    // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
    expect(handleChange.mock.calls[0][0]).toBe(true)
    handleChange.mockClear()
    expect(wrapper.render()).toMatchSnapshot()

    await movePlaceholder(-150)
    expect(handleChange.mock.calls.length).toBe(0)
    expect(wrapper.render()).toMatchSnapshot()
  })

  it('support offsetBottom', async () => {
    const handleChange = jest.fn()
    const wrapper = mount(<AffixMock offsetBottom={0} onChange={handleChange} />)

    await wait(20)
    expect(handleChange.mock.calls.length).toBe(0)
    expect(wrapper.render()).toMatchSnapshot()

    await movePlaceholder(50)
    expect(handleChange.mock.calls.length).toBe(0)
    expect(wrapper.render()).toMatchSnapshot()

    await movePlaceholder(150)
    expect(handleChange.mock.calls.length).toBe(1)
    // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
    expect(handleChange.mock.calls[0][0]).toBe(true)
    handleChange.mockClear()
    expect(wrapper.render()).toMatchSnapshot()

    await movePlaceholder(0)
    expect(handleChange.mock.calls.length).toBe(1)
    // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
    expect(handleChange.mock.calls[0][0]).toBe(false)
    handleChange.mockClear()
    expect(wrapper.render()).toMatchSnapshot()

    await movePlaceholder(-50)
    expect(handleChange.mock.calls.length).toBe(0)
    expect(wrapper.render()).toMatchSnapshot()

    await movePlaceholder(-150)
    expect(handleChange.mock.calls.length).toBe(0)
    expect(wrapper.render()).toMatchSnapshot()
  })
})
