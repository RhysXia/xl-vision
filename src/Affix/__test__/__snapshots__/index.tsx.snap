// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Affix support offsetBottom 1`] = `
<div
  class="container"
>
  <div>
    <div
      class=""
    >
      <button
        class="btn"
        type="button"
      >
        btn
      </button>
    </div>
  </div>
</div>
`;

exports[`Affix support offsetBottom 2`] = `
<div
  class="container"
>
  <div>
    <div
      class=""
    >
      <button
        class="btn"
        type="button"
      >
        btn
      </button>
    </div>
  </div>
</div>
`;

exports[`Affix support offsetBottom 3`] = `
<div
  class="container"
>
  <div>
    <div
      aria-hidden="true"
      style="width: 10px; height: 10px;"
    />
    <div
      class="xl-affix"
      style="position: fixed; bottom: 668px; width: 10px; height: 10px;"
    >
      <button
        class="btn"
        type="button"
      >
        btn
      </button>
    </div>
  </div>
</div>
`;

exports[`Affix support offsetBottom 4`] = `
<div
  class="container"
>
  <div>
    <div
      class=""
      style=""
    >
      <button
        class="btn"
        type="button"
      >
        btn
      </button>
    </div>
  </div>
</div>
`;

exports[`Affix support offsetBottom 5`] = `
<div
  class="container"
>
  <div>
    <div
      class=""
      style=""
    >
      <button
        class="btn"
        type="button"
      >
        btn
      </button>
    </div>
  </div>
</div>
`;

exports[`Affix support offsetBottom 6`] = `
<div
  class="container"
>
  <div>
    <div
      class=""
      style=""
    >
      <button
        class="btn"
        type="button"
      >
        btn
      </button>
    </div>
  </div>
</div>
`;

exports[`Affix support offsetTop 1`] = `
<div
  class="container"
>
  <div>
    <div
      class=""
    >
      <button
        class="btn"
        type="button"
      >
        btn
      </button>
    </div>
  </div>
</div>
`;

exports[`Affix support offsetTop 2`] = `
<div
  class="container"
>
  <div>
    <div
      class=""
    >
      <button
        class="btn"
        type="button"
      >
        btn
      </button>
    </div>
  </div>
</div>
`;

exports[`Affix support offsetTop 3`] = `
<div
  class="container"
>
  <div>
    <div
      class=""
    >
      <button
        class="btn"
        type="button"
      >
        btn
      </button>
    </div>
  </div>
</div>
`;

exports[`Affix support offsetTop 4`] = `
<div
  class="container"
>
  <div>
    <div
      class=""
    >
      <button
        class="btn"
        type="button"
      >
        btn
      </button>
    </div>
  </div>
</div>
`;

exports[`Affix support offsetTop 5`] = `
<div
  class="container"
>
  <div>
    <div
      aria-hidden="true"
      style="width: 10px; height: 10px;"
    />
    <div
      class="xl-affix"
      style="position: fixed; top: 0px; width: 10px; height: 10px;"
    >
      <button
        class="btn"
        type="button"
      >
        btn
      </button>
    </div>
  </div>
</div>
`;

exports[`Affix support offsetTop 6`] = `
<div
  class="container"
>
  <div>
    <div
      aria-hidden="true"
      style="width: 10px; height: 10px;"
    />
    <div
      class="xl-affix"
      style="position: fixed; top: 0px; width: 10px; height: 10px;"
    >
      <button
        class="btn"
        type="button"
      >
        btn
      </button>
    </div>
  </div>
</div>
`;
